subsys/net/ip/ipv4_autoconf.c

```diff
@@ -0,0 +1,263 @@
+	net_pkt_set_iface(pkt, iface);
+	net_pkt_set_family(pkt, AF_INET);
+	net_pkt_set_ll_reserve(pkt, sizeof(struct net_eth_hdr));
+
+	eth = NET_ETH_HDR(pkt);
+	hdr = NET_ARP_HDR(pkt);
+	eth->type = htons(NET_ETH_PTYPE_ARP);
+
+	/* ethernet header */
+	memcpy(&eth->src.addr, net_if_get_link_addr(iface)->addr,
+	       sizeof(struct net_eth_addr));
+	memset(&eth->dst.addr, 0xFF, sizeof(struct net_eth_addr));
+
+	hdr->hwtype = htons(NET_ARP_HTYPE_ETH);
+	hdr->protocol = htons(NET_ETH_PTYPE_IP);
+	hdr->hwlen = sizeof(struct net_eth_addr);
+	hdr->protolen = sizeof(struct in_addr);
+	hdr->opcode = htons(NET_ARP_REQUEST);
+
+	memcpy(&hdr->src_hwaddr.addr, &eth->src.addr,
+	       sizeof(struct net_eth_addr));
+	memset(&hdr->dst_hwaddr.addr, 0x00, sizeof(struct net_eth_addr));
+
+	net_ipaddr_copy(&hdr->dst_ipaddr, &iface->config.ipv4auto.current_ip);
+	net_ipaddr_copy(&hdr->src_ipaddr, &iface->config.ipv4auto.requested_ip);
+
+	net_buf_add(frag, sizeof(struct net_arp_hdr));
+
+	return pkt;
+
+fail:
+	net_pkt_unref(pkt);
+	return NULL;
+}
+
+enum net_verdict net_ipv4_autoconf_input(struct net_if *iface,
+					 struct net_pkt *pkt)
+{
+	struct net_arp_hdr *arp_hdr;
+
+	if (net_pkt_get_len(pkt) < (sizeof(struct net_arp_hdr) -
+				    net_pkt_ll_reserve(pkt))) {
+		NET_DBG("Invalid ARP header (len %zu, min %zu bytes)",
+			net_pkt_get_len(pkt),
+			sizeof(struct net_arp_hdr) -
+			net_pkt_ll_reserve(pkt));
+		return NET_DROP;
+	}
+
+	arp_hdr = NET_ARP_HDR(pkt);
+
+	if (!net_ipv4_addr_cmp(&arp_hdr->dst_ipaddr,
+			       &iface->config.ipv4auto.requested_ip)) {
+		/* No conflict */
+		return NET_DROP;
+	}
+
+#if NET_DEBUG > 0
+	do {
+		NET_DBG("conflict detected from %s for %s",
+			net_sprint_ll_addr((uint8_t *)&arp_hdr->src_hwaddr,
+					   arp_hdr->hwlen),
+			net_sprint_ipv4_addr(&arp_hdr->dst_ipaddr));
+	} while (0);
+#endif
+
+	iface->config.ipv4auto.conflict_cnt++;
+
+	switch (iface->config.ipv4auto.state) {
+	case NET_IPV4_AUTOCONF_PROBE:
+		/* restart probing with renewed IP */
+		net_ipv4_autoconf_start(iface);
+		break;
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+	case NET_IPV4_AUTOCONF_CONFIGURED:
+		if (iface->config.ipv4auto.conflict_cnt == 1) {
+			/* defend IP */
+			send_announcement(iface);
+		} else {
+			/* unset host ip */
+			if (!net_if_ipv4_addr_rm(
+				    iface, &iface->config.ipv4auto.requested_ip)) {
+				NET_DBG("Failed to remove addr from iface");
+			}
+			/* restart probing after second conflict */
+			net_ipv4_autoconf_start(iface);
+		}
+
+		break;
+	default:
+		break;
+	}
+
+	return NET_DROP;
+}
+
+static void ipv4_autoconf_timeout(struct k_work *work)
+{
+	struct net_if *iface =
+		CONTAINER_OF(work, struct net_if, config.ipv4auto.timeout);
+
+	if (!iface) {
+		NET_DBG("Invalid iface");
+		return;
+	}
+
+	switch (iface->config.ipv4auto.state) {
+	case NET_IPV4_AUTOCONF_PROBE:
+		/* schedule next probe */
+		if (iface->config.ipv4auto.probe_cnt <=
+		    IPV4_AUTOCONF_PROBE_NUM - 1) {
+			send_probe(iface);
+			break;
+		}
+
```

Comment:
Can you add comment here if this is intentional that we pass through without a break. So simple /* passthrough */ is enough, the idea is that anyone reading the code does not need to start to investigate if there is break missing or not.